Add a editable SkinEditor Element List

C0D3-M4513R opened the pull request:
This pr will attempt to implement #21098 .

done:
 - [x] Add the initial list implementation
 - [x] Sync the list contents to the drawables onscreen
 - [x] Add Lists that can be minimised
 - [x] Allow items in the List to be selected from the editor.
 - [x] Allow items to be selected from the List.
 - [x] Fix naming of the elements in the editor list
 - [x] Add a drag and drop handler, or another way to move IDrawableListItem's
 - [x] Link the changes to elements onscreen
 - [x] Allow those changes to be saved and loaded to config
 - [x] ~Fix having to set a fixed height for the OsuScrollContainer inside DrawableList/RearrangeableListContainer~ I do not think, that this is possible to fix. Please correct me, if I'm wrong.
 - [x] Optional: Rework DrawableListItem to simplify Selection Logic
 - [x] Optional: Allow Nested lists. ~This would also require adding a way to group together items in the editor, since I need a representative item for the DrawableMinimizableList, if I wanted to add it to a DrawableList (due to how RearrangeableListContainer is structured).~ 
 - [x] When Switching Screens the names don't get instantly updated
 - [x] Add Tests
 - [x] Rework how Lists are minimised. I do not like the OsuCheckbox for this. Arrows might work, when looking at how the enum selector works.
 - [x] Test drag on nested lists.
 - [x] Fix the Static Height in DrawableList
 - [x] Raise some sort of event (which includes old, and new parent) when changing parents of elements
 - [x] Should `OnDragAction` continue to be customisable? - No, it shouldn't.

optional: 
 - Rework, how to highlight selected items in a list - still looks completly fine, and may be easier to read currently.
 - Add user creatable nested lists? - potentially not this pr

C0D3-M4513R committed:
Add a division for a Layer-Editor in the sidebar

This divides the Settings Sidebar in the SkinEditor, to allow adding a Layer Editor.
Functionality for the Layer Editor is still Pending

C0D3-M4513R committed:
Add a List of elements, currently visible in ScreenEditor

C0D3-M4513R committed:
Allow Lists to be displayed by themselves

C0D3-M4513R committed:
Containers can now be minimised

C0D3-M4513R committed:
Populate the LayerEditor on SkinEditor start

C0D3-M4513R committed:
Sync element Selected status with the List

C0D3-M4513R committed:
Merge branch 'master' into skin-editor-layers

C0D3-M4513R committed:
Rework Selection to work both ways

This also gets rid of needing to implement IEquatable.
The Selection logic in DrawableListItem got a little big. I need to refactor it, to be simpler.
But selection works for now.

peppy commented on osu.Game/Screens/Edit/List/DrawableContainer.cs:
```diff
@@ -0,0 +1,115 @@
+
+namespace osu.Game.Screens.Edit.List
+{
+    public class DrawableContainer<T> : CompositeDrawable, IDrawableListItem<T>
```
This class name needs consideration.... It just can't be this.

peppy commented on osu.Game/Screens/Edit/List/DrawableList.cs:
```diff
@@ -0,0 +1,146 @@
+
+namespace osu.Game.Screens.Edit.List
+{
+    public class DrawableList<T> : CompositeDrawable, IDrawableListItem<T>
```
Is this something like `RearrangeableListContainer`? You might want to check whether you can already lean on an existing implementation like that.

C0D3-M4513R committed:
Rework Selection to have less duplicated code.

Also Selection now only works, if the target type extends IStateful<SelectionState>

C0D3-M4513R committed:
Revert "Rework Selection to have less duplicated code."

This reverts commit f5efbda2e29373ecd180df284e559ad3a3668fb2.

C0D3-M4513R committed:
Make SelectAll assignable from IDrawableListItem

Also Fixes Multi-Select being wierd

C0D3-M4513R committed:
Rename DrawableContainer to DrawableMinimisableList

C0D3-M4513R committed:
Use RearrangeableListContainer/RearrangeableListItem

C0D3-M4513R committed:
Merge branch 'master' into skin-editor-layers

C0D3-M4513R committed:
Fix Names in the Layer Editor

C0D3-M4513R committed:
Shift items, according to the Sidebar

C0D3-M4513R committed:
Allow Depth info to be saved and loaded.

C0D3-M4513R committed:
Fix skinning Items persisting Screen transitions

C0D3-M4513R committed:
Fix wrong Names getting displayed on scene change

C0D3-M4513R committed:
Merge branch 'master' into skin-editor-layers

C0D3-M4513R committed:
Rename IRearrangableListItem to IRearrangableDrawableListItem

IRearrangableListItem could have been confused, for the interface backing RearrangeableListItem.

C0D3-M4513R committed:
Merge branch 'master' into skin-editor-layers

C0D3-M4513R committed:
Merge branch 'master' into skin-editor-layers

peppy commented:
I think the general direction is good here, but it's very hard to review this change as it stands.

Here's a few pointers:
- How are you testing this? I'd hope to see some kind of test scene coverage to show how this interacts when there's more than one layer (and potentially nested layers).
- You mentioned having to give scroll container a height: this is correct. You probably want to make it `RelativeSizeAxes = Axes.Y` to have it expand to the containing grid container.
- I get that the checkbox to expand the list is temporary, but it's going to have to be replaced to get this in because it looks super weird at the moment. Maybe just use a simple `SpriteIcon` `+` and `-` (from fontawesome) for now?

peppy committed:
Fix inspections

peppy commented on osu.Game/Screens/Edit/List/DrawableList.cs:
```diff
@@ -0,0 +1,177 @@
+        protected override bool OnClick(ClickEvent e)
+        {
+            if (e.Button == MouseButton.Left)
+                SelectAll(SelectionState.NotSelected);
```
This reads super weird. "select all in a not selected state"?

peppy commented on osu.Game/Screens/Edit/List/DrawableList.cs:
```diff
@@ -0,0 +1,177 @@
+            return e.Button == MouseButton.Left;
+        }
+
+        public void AddRange(IEnumerable<T>? drawables)
+        {
+            if (drawables is null) return;
+
+            foreach (T drawable in drawables)
+            {
+                addInternal(drawable);
+            }
+
+            OnDragAction();
+        }
```
All this stuff seems too custom. I'd expect the list to just be a `Container` and handle things that way.

peppy commented on osu.Game/Screens/Edit/List/DrawableList.cs:
```diff
@@ -0,0 +1,177 @@
+            return false;
+        }
+
+        public bool Select(T drawable, bool value = true)
+        {
+            if (!ItemMap.ContainsKey(drawable)) return false;
+
+            return select(ItemMap[drawable], value);
+        }
+
+        public virtual void Select(bool value)
+        {
```
These overrides make no sense. One takes a drawable and one just..  takes a bool?

peppy commented on osu.Game/Screens/Edit/List/IDrawableListItem.cs:
```diff
@@ -0,0 +1,55 @@
+        /// <param name="value">if this List Item should be selected or not</param>
+        public void Select(bool value);
+
+        public void SelectableOnStateChanged(SelectionState obj)
+        {
+            switch (obj)
+            {
+                case SelectionState.Selected:
+                    SelectInternal(true);
+                    return;
+
+                case SelectionState.NotSelected:
+                    SelectInternal(false);
+                    return;
+            }
+        }
+
+        public LocalisableString GetDefaultText(Drawable target)
+        {
+            Logger.Log("GetDefaultText with" + target + " target.");
+            return target.Name.Equals(string.Empty) ? (target.GetType().DeclaringType ?? target.GetType()).Name : target.Name;
+        }
```
We generally avoid interface level default methods.

peppy reviewed (changes requested):
Very brief pass over code quality

peppy commented:
The general code quality here isn't up to scratch with the rest of the project. I think it may pay to read around some of the other classes we have and get a bit more used to the code base. I think this is going to be hard to move forward at this rate.

C0D3-M4513R committed:
Remove `IDrawableListItem.SelectableOnStateChanged`
Change `IDrawableListItem.SelectAll` to take bool instead of `SelectionState`

C0D3-M4513R committed:
Allow to select Multiple items in the DrawableList, when Shift if pressed.

C0D3-M4513R committed:
Don't unselect everything, if the user missclicks onto the DrawableList

C0D3-M4513R committed:
Remove SelectAll and replace it with ApplyAll

C0D3-M4513R committed:
Make `IDrawableListItem<T>.GetDefaultText` static

C0D3-M4513R committed:
Change how Actions Sent to `ApplyAll` apply

C0D3-M4513R committed:
Implement the Interfaces that Container<T> implements for DrawableList<T>

C0D3-M4513R committed:
Add a region for IDrawableListItem<T> implementation in DrawableList<T>

C0D3-M4513R committed:
Fix select in DrawableList<T> not selecting every IDrawableListItem<T>

C0D3-M4513R committed:
Add a [IDrawbable]RepresentedItem to IDrawableListItem

Change DrawableList to allow adding (sole) IDrawbableRepresentedItems or a IRearrangableDrawableListItem. Adding a IDrawableListItem to a DrawableList is not supported, and will result in a NullReferenceException.
This simultaneously allows me to get rid of the Various Container Methods.

C0D3-M4513R committed:
Fix OFSG001 and Code Quality: Make classes partial

C0D3-M4513R committed:
Fix NVIKA error

C0D3-M4513R committed:
Split SetItemDepth out of OnDragAction

C0D3-M4513R committed:
Add Tests for DrawableList

C0D3-M4513R committed:
Remove unnecessary use directives and add partial in TestSceneList

C0D3-M4513R committed:
Merge branch 'master' into skin-editor-layers

C0D3-M4513R committed:
Fix header and indentation in TestSceneList

C0D3-M4513R committed:
Fix NullReferenceException on test TearDown

C0D3-M4513R committed:
Fix InvalidThreadForMutationException in tests

C0D3-M4513R committed:
Add covariance to IDrawableListRepresetedItem

C0D3-M4513R committed:
Use a sprite for expanding a DrawableMinimisableList

C0D3-M4513R committed:
Rework Selection

C0D3-M4513R committed:
Rework List Show/Hide in DrawableMinimisableList

C0D3-M4513R committed:
Fix test nullable issue

C0D3-M4513R committed:
Fix test nullable issue

C0D3-M4513R committed:
Fix test nullable issue

C0D3-M4513R committed:
disable #nullable disable in SkinEditor and SkinBlueprintContainer

C0D3-M4513R committed:
Move variable init inline and out of load, for better nullable accuracy

C0D3-M4513R committed:
Fix LayerSidebarList being added to the Drawable Tree, after Disposed

C0D3-M4513R committed:
Add a tests for DrawableMinimisableList

C0D3-M4513R committed:
Remove unnecessary use directive

C0D3-M4513R committed:
Make SkinEditor.targetScreen variable Nullable

It is not given that targetScreen is set by a UpdateTargetScreen call, before it is used by e.g. Save() or in general by the availableTargets variable.

C0D3-M4513R committed:
Merge branch 'master' into skin-editor-layers

peppy commented:
This is starting to feel quite usable.

Immediate note is that when you click a top level container and it selects all content, the item's visible selection state doesn't update after deselecting one (or all) of the elements:


https://user-images.githubusercontent.com/191335/204985875-c41b6c8d-0be7-4812-bcd3-61636b545953.mp4

C0D3-M4513R committed:
Add Selection events to IRearrangableDrawableListItem
Fixes the representedListItem of a DrawableMinimisableList not being in the correct state after the selection state of it's children were altered

C0D3-M4513R committed:
Rework how items in Lists are added

This allows nested lists to work correctly, and imo makes the DrawableList.CreateDrawable method way easier to implement.

C0D3-M4513R committed:
Fix Tests and make DrawableMinimisableList.List not nullable

C0D3-M4513R committed:
Fix TestSceneList?

C0D3-M4513R committed:
Fix code style error IDE1006 - sigh

C0D3-M4513R committed:
Fix CodeFactor complaining

C0D3-M4513R committed:
Fix code style error IDE1006 again
Because I'm stupid and didn't rename the variable using refactoring.

C0D3-M4513R committed:
Fix NVIKA errors

C0D3-M4513R committed:
Merge branch 'master' into skin-editor-layers

C0D3-M4513R committed:
Merge branch 'master' into skin-editor-layers

C0D3-M4513R committed:
Make RearrangeableListContainer.ScrollContainer mostly ineffective

C0D3-M4513R committed:
Move commonly used properties into a shared DrawableListProperties instance

Properties that were supposed to be shared by all elements in the List previously had all elements being passed manually.
This is now not the case anymore. There should conceptually only be a single `DrawableListProperties` per List, not counting nested lists.

C0D3-M4513R committed:
Fix IDE1006 in DrawableList

C0D3-M4513R committed:
Initial revision of RearrangeableListContainer reimplementation

This is mostly a copy from RearrangeableListContainer and RearrangeableListItem, as a base to change the implementation and

C0D3-M4513R committed:
Initial support for cross list Item moves

This is the initial commit allowing for a item to be dragged into and out of Lists. Implementation details may change.

C0D3-M4513R committed:
Fix existing tests

C0D3-M4513R committed:
Remove RepresentedListItem in DrawableMinimisableList

C0D3-M4513R committed:
Prepare to add new tests

C0D3-M4513R committed:
Do a second pass over DrawableList.updateArrangement

This touches on general CodeQuality of commit 3e41e1c76befb3e8181aa4ca561a6be5ce62c983, and reworks a bit of stuff so it feels better.

This moves a part of DrawableList.updateArrangement into DrawableList.checkArrangement.
If no change needs to be made to the current layout or if the dragged element has been moved out of the list, we return a empty optional.
If the currentlyDraggedItem needs to be moved, a Value is returned.

This also removes the preprocessing statements that were added.

C0D3-M4513R committed:
Fix IDE0055 in TestSceneList

C0D3-M4513R committed:
Specify explicit test order

because without it, some tests failed test?
Rerunning failed tests would make all of them green.
Rerunning all tests of TestSceneList and TestSceneListMinimisable would then yield some failed tests again.
This should make it, so tests reliably pass.

C0D3-M4513R committed:
Remove unessesary variable.

Thanks NVika!

C0D3-M4513R committed:
Add ItemDraggedOut and ItemDraggedIn events

This commit prepares for user creatable lists (which I really do not wanna do)

C0D3-M4513R committed:
Merge branch 'master' into skin-editor-layers

C0D3-M4513R committed:
Fix SkinBlueprints not respecting Depth

Feodor0090 commented:
Something is wrong here.

Add something (default acc used) and something else (black box used)
![изображение](https://user-images.githubusercontent.com/53872073/206814045-c9421161-c384-4470-a6dc-31bc9c4ecfc5.png)

Black box is above everything, and it's in the bottom of list.

Then try to reorder it:
![изображение](https://user-images.githubusercontent.com/53872073/206814145-0b50f65d-60ce-430c-a6a0-dc9550c7d1ac.png)
![изображение](https://user-images.githubusercontent.com/53872073/206814162-83d4cb2a-1e14-40f7-a8b8-cc36fe23a653.png)

Now item in the bottom of list is under everything.

I checked "default" elements, they are ordered as "highest in list is the top in container":
![изображение](https://user-images.githubusercontent.com/53872073/206814456-f5a0b1f0-2fe5-4c50-8d3c-ff64a0ef5f3c.png)

Not sure which element should be above everything - top one or bottom one. Ignoring this, the only todo here is to add new elements in list on proper position (with current ordering to head, not to the end).

Blueprint ordering issue seems to be fixed by last commit.

C0D3-M4513R committed:
Fix newly added SkinElements had the wrong Depth

When newly adding items, the depth of the items would not be set Properly.

C0D3-M4513R committed:
Fix NVIKA "Redundant lambda signature parentheses"

C0D3-M4513R committed:
Fix Depth check

item more up the list should have a lower Depth, to be infront.

Feodor0090 commented:
Now new things are under everything. Aside from subjective point that this is weird, it changes behaviour (before 2 last commits new ones were on top). You should probably fix this.
![изображение](https://user-images.githubusercontent.com/53872073/206876510-2401ea6a-4f70-4f61-a8c7-f3c46348bde5.png)

Feodor0090 commented:
Just noticed more weird thing: notice how elements are flipping their order each gameplay enter.

https://user-images.githubusercontent.com/53872073/206876770-2c2aab1c-41dd-46b9-a87f-9a9a8f89cdc7.mp4

C0D3-M4513R committed:
Explicitly specify an order when adding elements

Before it could happen, that you could reload the same skin, and then the items are suddenly in a different order.
This should make that nearly impossible due to the following things:
 - When an item is added, it is always added directly before any items with the same depth
 - That is repeated for all items that are added (resulting in O(n^2))
 - After the item was added, the OnDragAction should fix the item Depths.

Feodor0090 commented:
Could not find any more ways to break this yet.

Two more nitpicks:
1) margins:
![изображение](https://user-images.githubusercontent.com/53872073/206881473-c73b9953-f3b7-4b4a-aee4-fe109770f346.png)
2) add something and a black box. Try to drag black box in the list. Notice that slider and checkbox flickers each mouse movement. Seems that the settings panel is recreated each drag. Would be nice not to do it.

C0D3-M4513R committed:
Fix adding items in Tests

C0D3-M4513R committed:
Make some lambdas static, where possible

C0D3-M4513R committed:
Apply Margin between `DrawableMinimisableList` Head and Content

C0D3-M4513R committed:
More static lambdas

C0D3-M4513R committed:
Make `OnDragAction` not customisable anymore

C0D3-M4513R committed:
Revert "Fix SkinBlueprints not respecting Depth"

This reverts commit 32d99015

C0D3-M4513R committed:
Fix SkinBlueprints selection being wonky sometimes

So it turns out, that I have been an complete idiot, and didn't actually set the Depths of items that were moved.

C0D3-M4513R committed:
The Item order was accidentally changed, by the revert

C0D3-M4513R committed:
Merge branch 'master' into skin-editor-layers

C0D3-M4513R committed:
Merge branch 'master' into skin-editor-layers

C0D3-M4513R committed:
Something about event nullability changed

This makes this pr compile again hopefully.

C0D3-M4513R committed:
Merge branch 'master' into skin-editor-layers

peppy committed:
Merge branch 'master' into skin-editor-layers

peppy commented:
This PR is unfortunately in a very hard place. The code style and sheer size of the PR (lines of code wise) makes it very hard to even start reviewing it. On the flip side, this *is* a component that we need for the skin editor, to both support depth adjustments and also in the future to support moving components between layers.

I don't really have any more to say for now, just thought I'd mention this.

C0D3-M4513R committed:
Merge branch 'master' into skin-editor-layers

# Conflicts:
#	osu.Game/Extensions/DrawableExtensions.cs
#	osu.Game/Overlays/SkinEditor/SkinBlueprintContainer.cs
#	osu.Game/Overlays/SkinEditor/SkinEditor.cs
#	osu.Game/Skinning/SerialisedDrawableInfo.cs

C0D3-M4513R committed:
Merge main, fix resulting errors

C0D3-M4513R committed:
Merge main, fix resulting errors
